### ISSUE TITLE
feature: MQTT Node: support content-type in v5 to preparse payload

### ISSUE THREAD

**Issue by knolleary** [node-red/node-red#3421]
### Description

Add to the existing parse options, but make use of the type hint v5 can provide.

**Comment by dceejay** [comment]
v5 has both a Payload Format indicator and a Content Type - should this issue use both as hints ?

As the Content type can be any Mime type - do you expect to parse them all or just a subset ?  Should it match what he http nodes try to do ?

**Comment by knolleary** [comment]
Short answer yes - there's plenty to explore here and resolve.

Not sure what 'parse any mime type' really means - because the output is either going to be an Object, String or Buffer. If it is a common format we know what to do with, then we should the right thing.

**Comment by Steve-Mcl** [comment]
@dceejay @knolleary I have given this some thought experiments & would appreciate feedback and ideas please...

## Content Type & Payload Format Indicator

### TL;DR

#### Step 1
* `Auto` mode, when accompanied with a Payload Format Indicator should
    * when `false` (i.e. undetermined byte payload), should simply pass the mqtt payload packet (a buffer by design) (no need to call the potentially expensive `isUtf8` check function)
     * when `true` (i.e. utf8 payload), should simply convert to string (no need to call the potentially expensive `isUtf8` check function)
  
#### Step 2 
* `Auto` mode, when accompanied with a recognised MIME type (TBD), should parse the incoming MQTT packet buffer into a string, JS Object or Buffer.


### Initial Thoughts
IMO, 
* Content Type is something more useful to the receiver of data (downstream nodes) so MIME types like `XML` or `msdoc` etc should be passed as delivered for downstream forking and conversions.
* Payload Format Indicator should be used to "assist" the auto determination of the output, potentially avoiding expensive tests on `isUtf8`

### However, that said
If the `MQTT In` nodes "Output" option is set for "Auto" we _should_ consider common mime types (TBD) and the Payload Packet Indicator to assist in what checks and conversions are performed to auto determine the output to be one of JS Object, String or Buffer. 

### Proposed Logic

The some of the obvious Output Modes / combinations  (buffer, to String, to Base64 String) entries are skipped for brevity



Output Mode | V5 Content Type | V5 Payload Format Indicator | Incoming Data | Result | Comments
-- | -- | -- | -- | -- | --
Auto | empty | empty | utf8 string | utf8 string | same as now
Auto | empty | empty | non utf8 | buffer | same as now
Parse JSON | empty | empty | any | JS Object | same as now
to String | empty | empty | any | utf8 string | same as now
to Buffer | empty | empty | any | buffer | same as now
to Base64 String | empty | empty | any | utf8 string | same as now
Auto | empty | utf8 | any | utf8 string | *3
Auto | empty | byte | any | buffer | *2
Parse JSON | empty | utf8 | any | JS Object | *1
to String | empty | utf8 | any | utf8 string | *3
to Buffer | empty | byte | any | buffer | *2
to Base64 String | empty | utf8 | any | utf8 string | *3
Auto | application/json | any | any | JS Object | *1
Auto | text/html | any | any | string | *3
Auto | text/plain | any | any | string | *3
Auto | text/html | any | any | string | *3
Auto | text/richtext | any | any | string | *3
Auto | application/octet-stream | any | any | buffer | *2
Auto | application/pdf | any | any | buffer | *2
Auto | application/x-gtar | any | any | buffer | *2
Auto | application/x-gzip | any | any | buffer | *2
Auto | application/x-tar | any | any | buffer | *2
Auto | application/zip | any | any | buffer | *2
Auto | audio/aac | any | any | buffer | *2
Auto | audio/x-aiff | any | any | buffer | *2
Auto | audio/basic | any | any | buffer | *2
Auto | audio/x-midi | any | any | buffer | *2
Auto | audio/x-wav | any | any | buffer | *2
Auto | image/bmp | any | any | buffer | *2
Auto | image/gif | any | any | buffer | *2
Auto | image/jpeg | any | any | buffer | *2
Auto | image/tiff | any | any | buffer | *2
Auto | image/x-xbitmap | any | any | buffer | *2






Remark | Detail
-- | --
*1 | Skip expensive call to `isUtf8` & go ahead `try` to convert to JS Object
*2 | Make no attempt to parse incoming MQTT buffer - just output it. Avoids expensive call to `isUtf8`
*3 | Make no attempt to parse incoming MQTT buffer - just call `toString`.


### NOTES
* When MQTT V5 was merged, the payload format indicator was commented out/excluded on the MQTT Out node (though it can still be set via `msg.payloadFormatIndicator`) - we may want to consider adding this back in?

**Comment by dceejay** [comment]
sorry to be a pain... so what are the proposed rules for v3 connections ?
Ok maybe not part of this Issue as it is v5  - but presumably you are looking at that as well and I can't see a cross ref to another issue.

**Comment by Steve-Mcl** [comment]
Hi @dceejay 

No pain at all mate. Lets do what is best for node-red :) 

So, I have not yet gave a great deal of thought to improvements for V3 - do you have any suggestions? 

In fact, do you have any suggestions regarding V5 also?  I seem to remember reading somewhere about the auto mode should attempt to parse JSON directly to a JS object?  @knolleary was this mentioned?

As it stands, I have left Auto mode doing String or Buffer (for compatibility) but I think Auto should be able to also do JSON --> object but it might cause users some (small, easily fixable) problems.  Would this be acceptable for a major version release considering the benefits?

PS: In my tests, using the v5 payload format indicator (and/or the contentType) to skip the `isUtf8` call, there are significant gains in the reception of large text and especially JSON (since JSON is naturally UTF8 & with the payload format indicator set true, we skip the  `isUtf8` call).  On a reasonably sized JSON (3000 objects, approx 3MB total) there is a saving (on my i7 machine) of between 11~15ms per call.

**Comment by dceejay** [comment]
Indeed your thoughts re auto mode for v3 echo  mine - I think auto should mean exactly that - it should try to parse to JSON first - then text - then buffer... and yes it is a breaking change - but I think it is within bounds for v3. (and indeed if performance is an issue then don't use auto mode ;-)

If you are hinted at the type with v5 then yes no need to test - as long as we then indicate errors when they aren't of that type etc...

**Comment by Steve-Mcl** [comment]
@dceejay good stuff. Let's see what Nick says.

**Comment by Steve-Mcl** [comment]
@dceejay I have updated the code to auto parse a string to JSON (so long as it is UTF8)

https://github.com/Steve-Mcl/node-red/blob/ce67737cc95439c9de9ce4114eac4db0c47252ec/packages/node_modules/%40node-red/nodes/core/network/10-mqtt.js#L270

I will raise PR soon.
